### PR TITLE
Migrate to React Router V6

### LINKED ISSUE #249: Switch to react-router 6
RR6 is currently in alpha but its design is relevant to how we would like to structure the app.

### PR DESCRIPTION
**Pull request by benjdlambert**
Bored at the weekend so decided to have a punt at starting some of the work to move over to React-Router v6. Still work in progress but the catalog page seems to work with subrouting to the component page.

Following this migration guide: https://github.com/ReactTraining/react-router/blob/f59ee5488bc343cf3c957b7e0cc395ef5eb572d2/docs/advanced-guides/migrating-5-to-6.md#relative-routes-and-links

#### :heavy_check_mark: Checklist

- [x] All tests are passing `yarn test`
- [ ] Screenshots attached (for UI changes)
- [ ] Relevant documentation updated
- [x] Prettier run on changed files
- [x] Tests added for new functionality
- [x] Regression tests added for bug fixes

### CODE REVIEW

File: plugins/lighthouse/src/components/CreateAudit/index.test.tsx [hunk 1]
```diff
@@ -140,7 +140,7 @@ describe('CreateAudit', () => {
 
       await wait(() => expect(rendered.getByLabelText(/URL/)).toBeEnabled());
 
-      expect(useHistory().push).toHaveBeenCalledWith('/lighthouse');
+      expect(useNavigate().push).toHaveBeenCalledWith('/lighthouse');
```
Comment (Rugvip) [inline plugins/lighthouse/src/components/CreateAudit/index.test.tsx:143]: is this supposed to still be push?

File: packages/core/src/layout/Sidebar/Items.tsx [hunk 2]
```diff
@@ -139,8 +139,7 @@ export const SidebarItem: FC<SidebarItemProps> = ({
       <NavLink
         className={clsx(classes.root, classes.closed)}
         activeClassName={classes.selected}
-        isActive={match => Boolean(match && !disableSelected)}
-        exact
+        end
```
Comment (Rugvip) [inline packages/core/src/layout/Sidebar/Items.tsx:143]: Do you know what's up with this API? Looks pretty weird :grin:

**Review by Rugvip — COMMENTED** [review]
:shipit:

File: plugins/catalog/src/components/ComponentPage/ComponentPage.tsx [hunk 3]
```diff
@@ -48,17 +49,19 @@ type ComponentPageProps = {
...
 };
 
-const ComponentPage: FC<ComponentPageProps> = ({ match, history }) => {
+const ComponentPage: FC<ComponentPageProps> = () => {
   const [confirmationDialogOpen, setConfirmationDialogOpen] = useState(false);
   const [removingPending, setRemovingPending] = useState(false);
   const showRemovalDialog = () => setConfirmationDialogOpen(true);
   const hideRemovalDialog = () => setConfirmationDialogOpen(false);
-  const componentName = match.params.name;
+  const params = useParams() as { name: string };
```
[the remark this answers is not in the document]
Comment (benjdlambert) [reply plugins/catalog/src/components/ComponentPage/ComponentPage.tsx:57]: I tried that but it didn’t work, maybe there’s a different way that’s undocumented, but using `<{ name: string }>` with the hook isn’t supported

File: packages/core/src/layout/Sidebar/Items.tsx (hunk 2 above)
Comment (benjdlambert) [reply packages/core/src/layout/Sidebar/Items.tsx:143]: Haha no idea. It wasn’t available until the latest alpha. Strange name.

File: plugins/lighthouse/src/components/CreateAudit/index.test.tsx (hunk 1 above)
Comment (benjdlambert) [reply plugins/lighthouse/src/components/CreateAudit/index.test.tsx:143]: Nope, I just haven’t got round to fixing the tests yet haha. This was find and replace 😂

File: plugins/lighthouse/src/components/CreateAudit/index.tsx [hunk 4]
```diff
@@ -113,7 +113,7 @@ const CreateAudit: FC<{}> = () => {
           <Grid item xs={12} sm={6}>
             <InfoCard>
               <form
-                onSubmit={(ev) => {
+                onSubmit={ev => {
```
Comment (benjdlambert) [inline plugins/lighthouse/src/components/CreateAudit/index.tsx:116]: Not sure if this is desired? Seems like prettier is configured badly in some places if so

Comment (freben) [reply plugins/lighthouse/src/components/CreateAudit/index.tsx:116]: This is how we set it afaik

Comment (benjdlambert) [reply plugins/lighthouse/src/components/CreateAudit/index.tsx:116]: @freben sorry let me rephrase, should we not lint at build time still?

I mean I like the fact that it doesn't complain for stuff that you haven't changed, but it just makes me worried only the most maintained files will have the latest lint standards. And when you have a lint rule that prefers something like `import {...} from '@backstage/core'` over `../../../packages/core` it can make simple renaming tasks a pain.

File: packages/core/src/components/Link/Link.stories.tsx [hunk 5]
```diff
@@ -73,7 +73,7 @@ export const PassProps = () => {
       <Link
         to={routeRef.path}
         /** react-router-dom related prop */
-        component={RouterNavLink}
+        as={RouterNavLink}
```
Comment (Rugvip) [inline packages/core/src/components/Link/Link.stories.tsx:76]: Is this a prop that is passed to our link component, through to the MUI link component, and finally to the react-router link component, to render with another link component?

File: plugins/circleci/src/components/App.tsx [hunk 6]
```diff
@@ -24,14 +24,13 @@ export const App = () => {
   return (
     <AppStateProvider>
       <>
-        <Switch>
-          <Route path="/circleci" exact component={BuildsPage} />
+        <Routes>
+          <Route path="/circleci" element={<BuildsPage />} />
```
Comment (Rugvip) [inline plugins/circleci/src/components/App.tsx:28]: So now we can refactor this to just be `path="/"` right?

**Review by Rugvip — APPROVED** [review]

File: packages/core/src/components/Link/Link.stories.tsx (hunk 5 above)
Comment (benjdlambert) [reply packages/core/src/components/Link/Link.stories.tsx:76]: Ah, sorry. Yeah this is why I got all confused. This should be reverted back to `component` again. Will fix now.

File: plugins/circleci/src/components/App.tsx (hunk 6 above)
Comment (benjdlambert) [reply plugins/circleci/src/components/App.tsx:28]: Yes, if the encapsulating `Routes` has `basename` of `/circleci` then the `Route` can just use `/`
